java/cdk-custom-resources/src/main/java/sleeper/cdk/custom/AutoStopEmrServerlessApplicationLambda.java

```diff
@@ -0,0 +1,140 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package sleeper.cdk.custom;
+
+import com.amazonaws.services.lambda.runtime.Context;
+import com.amazonaws.services.lambda.runtime.events.CloudFormationCustomResourceEvent;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import software.amazon.awssdk.services.emrserverless.EmrServerlessClient;
+import software.amazon.awssdk.services.emrserverless.model.ApplicationState;
+import software.amazon.awssdk.services.emrserverless.model.ApplicationSummary;
+import software.amazon.awssdk.services.emrserverless.model.JobRunState;
+import software.amazon.awssdk.services.emrserverless.model.JobRunSummary;
+
+import sleeper.core.util.PollWithRetries;
+
+import java.time.Duration;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * Delete an EMR Serverless application.
+ */
+public class AutoStopEmrServerlessApplicationLambda {
+    public static final Logger LOGGER = LoggerFactory.getLogger(AutoStopEmrServerlessApplicationLambda.class);
+
+    private final PollWithRetries poll = PollWithRetries
+            .intervalAndPollingTimeout(Duration.ofSeconds(30), Duration.ofMinutes(15));
+    private final EmrServerlessClient emrServerlessClient;
+    private String applicationPrefix;
+
+    public AutoStopEmrServerlessApplicationLambda() {
+        this(EmrServerlessClient.create());
+    }
+
+    public AutoStopEmrServerlessApplicationLambda(EmrServerlessClient emrServerlessClient) {
+        this.emrServerlessClient = emrServerlessClient;
+    }
+
+    /**
+     * Handles an event triggered by CloudFormation.
+     *
+     * @param event   the event to handle
+     * @param context the context
+     */
+    public void handleEvent(
+            CloudFormationCustomResourceEvent event, Context context) throws InterruptedException {
+
+        Map<String, Object> resourceProperties = event.getResourceProperties();
+        this.applicationPrefix = "sleeper-" + (String) resourceProperties.get("instanceId");
+
+        switch (event.getRequestType()) {
+            case "Create":
+                break;
+            case "Update":
+                break;
+            case "Delete":
+                stopEmrServerlessApplication(emrServerlessClient);
+                break;
+            default:
+                throw new IllegalArgumentException("Invalid request type: " + event.getRequestType());
+        }
+    }
+
+    private void stopEmrServerlessApplication(EmrServerlessClient emrServerless) throws InterruptedException {
+        this.run();
+    }
+
+    private void run() throws InterruptedException {
+        List<ApplicationSummary> applications = listActiveApplications();
```

Comment:
I don't think we need to list the applications. Looking in EmrServerlessBulkImportStack, we create one application and we have its application ID there. We can pass in the application ID as a resource property instead of the instance ID.